### PR TITLE
Allow GPIO 9 and 10 for waveform generation

### PR DESCRIPTION
**Pull request by earlephilhower**
While most ESP8266 modules use quad-io mode for their SPI flash ROM,
there are some which only use dual-io mode.  Allow the unused pins
(GPIO 9 and 10) to have waveforms generated on them.  Should the user try
this on a quad-io mode board, expect very bad things to happen.

### CODE REVIEW

**Comment by d-a-v** [comment]
Correct me if I'm wrong:
- all DIO boards can use GPIO9&10, unlike QIO boards ?
- DIO can be used on QIO boards, does that mean that we can use GPIO9&10 on every boards if they are configured in DIO mode ? 

The board generator knows them well and could define -DDIO=1 for all the DIO-only ones.
It's not that I'm determined, I would if I had more time :) That could be a new issue.

**Comment by earlephilhower** [comment]
I believe the flash chip on the module needs a strapping change to set it to x2 or x4 interface, so I don't think you're quite right on the second portion.

On DIO boards GPIO9/10 (aka SDIO_DATA_3/4) are not used by the ICACHE to talk to program flash, so theoretically you can use them.  Whether or not they are brought out to a header (and defined as a Arduino-style D* pin) is up to the manufacturer.

It's possible, but for all that work you end up saving ~20 bytes of heap in the best case.  `pinMode` and `digitalWrite` would still be able to do bad things if applied to the SDIO pins, unless you also checked there.  Right now, for speed, I believe there is no checking at all.

**Comment by igrr** [comment]
GPIOs 9 and 11 can be used as GPIOs when flash is in DIO mode *and* these IOs are not connected to the flash chip. The latter is not true at least for the popular ESP-12 modules found on nodemcu boards and clones.
They can be used on ESP8285 though.

**Comment by earlephilhower** [comment]
> GPIOs 9 and 11 can ...

You meant "9 and 10", right?  GPIO11 is the CMD pin per the tech ref manual.

**Comment by igrr** [comment]
9 and 10, yes, sorry.

**Comment by devyte** [comment]
There was dicsussion about  pins 9 and 10 some time ago for the sonoff (I think), for using them as interrupt inputs. I believe there was some conditional code  done in the pin init code to init these pins. That should probably be connected to this.

**Comment by earlephilhower** [comment]
The Sonoff/GPIO9/10 IRQ thing was referenced in #3549, but according to #2257 it actually already works, @devyte.

I think that the Sonoff thing was just a programming error.  You need to set pinMode(xxx, INPUT) before the IRQ attach, most likely (and was not done in the MCVE-snippet in 3549), because that's not a state that can be safely set on anything but those 2xDIO devices.

**Comment by d-a-v** [comment]
Should we allow those pins for esp8285 only ?

**Comment by devyte** [comment]
That's sort of my point: it works only for 8285 and only by initing them as input, as in before the irq attach, which currently isn't done in the core. Also, thenattach is currently hidden from the user.
Should this be added as a board specific init?

**Comment by igrr** [comment]
Probably need to define new variant, and do this in initVariant. ESP8285 board can then specify this new variant.

**Comment by devyte** [comment]
@earlephilhower what I meant is [this function](https://github.com/esp8266/Arduino/blob/c33ef89f603b81b01a1e4c5cebd89be4fbfe7b29/cores/esp8266/core_esp8266_wiring_digital.c#L214), where the pins are inited on startup. The user doesn't have access to that function (i.e.: is hidden), and the init is currently hardcoded. What I mean is whether there should be an additional init there for pins 9 and 10, depending on whether it's an ESP8285 board.
I don't think that just checking whether flash mode is DIO is enough, because you can configure e.g.: an ESP12 for DIO, but the pins are still wired to the chip, and they must not be per @igrr .

**Comment by devyte** [comment]
I think that @d-a-v 's comment above is for the same reason. Then, we should enable waveform for those pins only for the ESP8285 boards as well.

**Comment by earlephilhower** [comment]
The most logical way to do it seems to be by making an esp8285 variant.  Can you guys look at it and see if it makes sense?  This is effectively the only board you're going to find in the wild using 2-pin mode, so it should not be a maintenance problem.

**Comment by earlephilhower** [comment]
Updated to fix the flash pin interface macro, and use that to hide pins 9/10 when in non-8285.  Should address all the issues, I think.

File: cores/esp8266/core_esp8266_waveform.c [hunk 1]
```diff
@@ -79,7 +79,11 @@ static Waveform waveform[] = {
   {0, 0, 1<<3, 0, 0, 0, 0, 0},
   {0, 0, 1<<4, 0, 0, 0, 0, 0},
   {0, 0, 1<<5, 0, 0, 0, 0, 0},
-  // GPIOS 6-11 not allowed, used for flash
+  // GPIOS 6-8 not allowed, used for flash
+  // GPIO 9 and 10 only allowed in 2-bit flash mode
+  {0, 0, 1<<9, 0, 0, 0, 0, 0},
```
[the remark this answers is not in the document]
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: Done with the fixed isFlashInterfacePin() macro.

**Comment by terminet85** [comment]
With 2.6.3, GPIO 10, only works as Digital Input Output, no PWM...

**Pull request merged**